pgrx/src/shmem_hash.rs

```diff
@@ -0,0 +1,250 @@
+//! Shared memory hash map implemented with Postgres' internal `HTAB`,
+//! which is used by other extensions like `pg_stat_statements`.
+use crate::{pg_sys, shmem::PgSharedMemoryInitialization, spinlock::*, PGRXSharedMemory};
+use once_cell::sync::OnceCell;
+use std::ffi::c_void;
+use uuid::Uuid;
+
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum ShmemHashMapError {
+    /// Hash table can't have more entries due to fixed allocation size.
+    HashTableFull,
+}
+
+#[derive(Copy, Clone, Debug)]
+struct ShmemHashMapInner {
+    htab: *mut pg_sys::HTAB,
+    elements: i64,
+}
+
+unsafe impl PGRXSharedMemory for ShmemHashMapInner {}
+unsafe impl Send for ShmemHashMapInner {}
+unsafe impl Sync for ShmemHashMapInner {}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug)]
+struct Key<K> {
+    // We copy it with std::ptr::copy, but we don't actually use the field
+    // in Rust, hence the warning.
+    #[allow(dead_code)]
+    key: K,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug)]
+struct Value<K, V> {
+    #[allow(dead_code)]
+    key: Key<K>,
+    value: V,
+}
+
+impl Default for ShmemHashMapInner {
+    fn default() -> Self {
+        Self { htab: std::ptr::null_mut(), elements: 0 }
+    }
+}
+
+/// A shared memory HashMap using Postgres' `HTAB`.
+/// This HashMap is used for `pg_stat_statements` and Postgres
+/// internals to store key/value pairs in shared memory.
+pub struct ShmemHashMap<K: Copy + Clone, V: Copy + Clone> {
+    /// HTAB protected by a SpinLock.
+    htab: OnceCell<PgSpinLock<ShmemHashMapInner>>,
+
+    /// Max size, allocated at server start.
+    size: i64,
+
+    // Markers for key/value types.
+    _phantom_key: std::marker::PhantomData<K>,
+    _phantom_value: std::marker::PhantomData<V>,
+}
+
+/// Compute the hash for the key and its pointer
+/// to pass to `pg_sys::hash_search_with_hash_value`.
+/// Lock on HTAB should be taken, although not strictly required I think.
+macro_rules! key {
+    ($key:expr, $htab:expr) => {{
+        let key = Key { key: $key };
+        let key_ptr: *const c_void = std::ptr::addr_of!(key) as *const Key<K> as *const c_void;
+        let hash_value = unsafe { pg_sys::get_hash_value($htab.htab, key_ptr) };
```

Comment:
Same comment re: "this has to document the preconditions of this function so we know why it's safe to call". The answer of "there are virtually none and this is an in-principle safe function as long as we call it when loaded in the Postgres context" is fine.